File: sigil-hs/src/Sigil/IO/Reader.hs
module Sigil.IO.Reader
  ( decodeSigilFile
  , readSigilFile
  ) where

import Data.Binary.Get
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import Data.Text.Encoding (decodeUtf8')
import Data.Word (Word8)

import Sigil.Core.Types
import Sigil.Core.Error
import Sigil.Core.Chunk
import Sigil.Codec.Pipeline (decompress)

magic :: ByteString
magic = BS.pack [0x89, 0x53, 0x47, 0x4C, 0x0D, 0x0A]

decodeSigilFile :: BL.ByteString -> Either SigilError (Header, Metadata, Image)
decodeSigilFile input = case runGetOrFail parser input of
  Left (_, _, err) -> Left (IoError err)
  Right (_, _, result) -> result
  where
    parser = do
      m <- getByteString 6
      if m /= magic
        then pure $ Left (InvalidMagic m)
        else do
          major <- getWord8
          minor <- getWord8
          if major /= 0 || minor /= 2
            then pure $ Left (UnsupportedVersion major minor)
            else do
              chunks <- readChunks
              parseChunks chunks

readChunks :: Get [Chunk]
readChunks = do
  tag <- getByteString 4
  len <- getWord32be
  payload <- getByteString (fromIntegral len)
  crcVal <- getWord32be
  case tagFromBytes tag of
    Left _err -> pure []
    Right t -> do
      let chunk = Chunk t payload crcVal
      if t == SEND
        then pure [chunk]
        else do
          rest <- readChunks
          pure (chunk : rest)

parseChunks :: [Chunk] -> Get (Either SigilError (Header, Metadata, Image))
parseChunks chunks = pure $ do
  -- Verify all CRCs
  mapM_ verifyChunk chunks
  -- Find SHDR
  shdr <- case filter (\c -> chunkTag c == SHDR) chunks of
    (c:_) -> Right c
    []    -> Left (MissingChunk "SHDR")
  hdr <- decodeHeader (chunkPayload shdr)
  -- Optional SMTA
  let meta = case filter (\c -> chunkTag c == SMTA) chunks of
        (c:_) -> case decodeMetadata (chunkPayload c) of
                   Right m -> m
                   Left _  -> emptyMetadata
        []    -> emptyMetadata
  -- Concatenate SDAT payloads
  let sdatPayload = BS.concat
        [ chunkPayload c | c <- chunks, chunkTag c == SDAT ]
  img <- decompress hdr sdatPayload
  Right (hdr, meta, img)

decodeHeader :: ByteString -> Either SigilError Header
decodeHeader bs = case runGetOrFail parser (BL.fromStrict bs) of
  Left (_, _, _err) -> Left TruncatedInput
  Right (_, _, hdr) -> hdr
  where
    parser = do
      w <- getWord32be
      h <- getWord32be
      cs <- getWord8
      bd <- getWord8
      p  <- getWord8
      pure $ do
        colorSp <- toColorSpace cs
        bitD    <- toBitDepth bd
        pred'   <- toPredictorId p
        when' (w == 0 || h == 0) $ Left (InvalidDimensions w h)
        Right (Header w h colorSp bitD pred')

    toColorSpace :: Word8 -> Either SigilError ColorSpace
    toColorSpace 0 = Right Grayscale
    toColorSpace 1 = Right GrayscaleAlpha
    toColorSpace 2 = Right RGB
    toColorSpace 3 = Right RGBA
    toColorSpace n = Left (InvalidColorSpace n)

    toBitDepth :: Word8 -> Either SigilError BitDepth
    toBitDepth 8  = Right Depth8
    toBitDepth 16 = Right Depth16
    toBitDepth n  = Left (InvalidBitDepth n)

    toPredictorId :: Word8 -> Either SigilError PredictorId
    toPredictorId n
      | n <= fromIntegral (fromEnum (maxBound :: PredictorId)) = Right (toEnum (fromIntegral n))
      | otherwise = Left (InvalidPredictor n)

    when' False _ = Right ()
    when' True e  = e

decodeMetadata :: ByteString -> Either SigilError Metadata
decodeMetadata bs = case runGetOrFail parser (BL.fromStrict bs) of
  Left _ -> Right emptyMetadata
  Right (_, _, entries) -> Right (Metadata entries)
  where
    parser = do
      empty <- isEmpty
      if empty then pure []
      else do
        kLen <- getWord16be
        kBs <- getByteString (fromIntegral kLen)
        vLen <- getWord32be
        vBs <- getByteString (fromIntegral vLen)
        case decodeUtf8' kBs of
          Left _  -> pure []
          Right k -> do
            rest <- parser
            pure ((k, vBs) : rest)

readSigilFile :: FilePath -> IO (Either SigilError (Header, Metadata, Image))
readSigilFile path = do
  bs <- BL.readFile path
  pure (decodeSigilFile bs)

File: sigil-rs/src/reader.rs
use crate::chunk::{parse_chunks, Tag};
use crate::error::SigilError;
use crate::pipeline::decompress;
use crate::types::{Header, ColorSpace, BitDepth, PredictorId};

const MAGIC: [u8; 6] = [0x89, 0x53, 0x47, 0x4C, 0x0D, 0x0A];

/// Parse an .sgl file and return the header + raw pixel data.
pub fn decode(data: &[u8]) -> Result<(Header, Vec<u8>), SigilError> {
    let header = read_header(data)?;

    // Skip magic (6) + version (2) = 8 bytes
    let chunks = parse_chunks(data, 8)?;

    // Concatenate SDAT payloads
    let mut sdat_payload = Vec::new();
    for chunk in &chunks {
        if chunk.tag == Tag::Sdat {
            sdat_payload.extend_from_slice(chunk.payload);
        }
    }

    let pixels = decompress(&header, &sdat_payload)?;
    Ok((header, pixels))
}

/// Read only the header without decoding pixel data.
pub fn read_header(data: &[u8]) -> Result<Header, SigilError> {
    // Validate magic
    if data.len() < 8 {
        return Err(SigilError::TruncatedInput);
    }
    if data[0..6] != MAGIC {
        return Err(SigilError::InvalidMagic);
    }

    // Validate version
    let major = data[6];
    let minor = data[7];
    if major != 0 || minor != 3 {
        return Err(SigilError::UnsupportedVersion { major, minor });
    }

    // Parse chunks to find SHDR
    let chunks = parse_chunks(data, 8)?;
    let shdr = chunks.iter().find(|c| c.tag == Tag::Shdr)
        .ok_or(SigilError::MissingChunk("SHDR"))?;

    parse_header(shdr.payload)
}

fn parse_header(payload: &[u8]) -> Result<Header, SigilError> {
    if payload.len() < 11 {
        return Err(SigilError::TruncatedInput);
    }

    let width = u32::from_be_bytes([payload[0], payload[1], payload[2], payload[3]]);
    let height = u32::from_be_bytes([payload[4], payload[5], payload[6], payload[7]]);

    if width == 0 || height == 0 {
        return Err(SigilError::InvalidDimensions(width, height));
    }

    let color_space = ColorSpace::from_byte(payload[8])
        .ok_or(SigilError::InvalidColorSpace(payload[8]))?;

    let bit_depth = BitDepth::from_byte(payload[9])
        .ok_or(SigilError::InvalidBitDepth(payload[9]))?;

    let predictor = PredictorId::from_byte(payload[10])
        .ok_or(SigilError::InvalidPredictor(payload[10]))?;

    Ok(Header { width, height, color_space, bit_depth, predictor })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_invalid_magic() {
        let data = [0x00; 100];
        assert_eq!(decode(&data).unwrap_err(), SigilError::InvalidMagic);
    }

    #[test]
    fn test_unsupported_version() {
        let mut data = [0u8; 100];
        data[0..6].copy_from_slice(&MAGIC);
        data[6] = 1; // major = 1
        data[7] = 0; // minor = 0
        assert_eq!(decode(&data).unwrap_err(), SigilError::UnsupportedVersion { major: 1, minor: 0 });
    }

    #[test]
    fn test_truncated() {
        assert_eq!(decode(&[0x89]).unwrap_err(), SigilError::TruncatedInput);
    }
}
